internal/pkg/service/common/etcdop/iterator/iterator.go

```diff
@@ -10,46 +11,175 @@ import (
 	"github.com/keboola/keboola-as-code/internal/pkg/utils/errors"
 )
 
-type Definition[R any] struct {
+const (
+	end = ""
+)
+
+type Definition[T any] struct {
 	config
 }
 
-type Iterator[R any] struct {
+type Iterator[T any] struct {
 	config
-	ctx    context.Context
-	client *etcd.Client
-	page   int
-	index  int
-	err    error
+	ctx          context.Context
+	client       *etcd.Client
+	err          error
+	start        string          // page start prefix
+	page         int             // page number, start from 1
+	lastIndex    int             // lastIndex in the page, 0 means empty
+	currentIndex int             // currentIndex in the page, start from 0
+	revision     int64           // revision of the all values, set on the first page
+	values       []*op.KeyValue  // values in the page
+	currentValue op.KeyValueT[T] // currentValue in the page, match currentIndex
 }
 
 func New[R any](start string, serde serde.Serde, opts ...Option) Definition[R] {
-	// Apply options
-	c := config{limit: DefaultLimit, start: start, serde: serde}
-	for _, o := range opts {
-		o(&c)
-	}
-	return Definition[R]{config: c}
+	return Definition[R]{config: newConfig(start, serde, opts)}
 }
 
+// Do converts iterator definition to the iterator.
 func (v Definition[T]) Do(ctx context.Context, client *etcd.Client) *Iterator[T] {
-	return &Iterator[T]{ctx: ctx, client: client, config: v.config, page: -1, index: -1}
+	return &Iterator[T]{ctx: ctx, client: client, config: v.config, start: v.config.prefix, page: 0, currentIndex: 0}
 }
 
+// Next returns true if there is a next value.
+// False is returned if there is no next value or an error occurred.
 func (v *Iterator[T]) Next() bool {
-	return false
+	select {
+	case <-v.ctx.Done():
+		// Stop iteration if the context is done
+		v.err = v.ctx.Err()
+		return false
+	default:
+		// Is there one more item?
+		if !v.nextItem() && !v.nextPage() {
+			return false
+		}
+
+		// Decode item
+		v.currentValue = op.KeyValueT[T]{KV: v.values[v.currentIndex]}
+		if err := v.serde.Decode(v.ctx, v.currentValue.KV, &v.currentValue.Value); err != nil {
+			v.err = errors.Errorf(`etcd iterator failed: cannot decode key "%s", page=%d, index=%d: %w`, v.currentValue.KV.Key, v.page, v.currentIndex, err)
+		}
+		return v.err == nil
+	}
```

Comment:
Next zaroven aj dekoduje a validuje hodnotu.

Dalo by sa to optimalizovat:
- nacitat viac stranok dopredu na pozadi
- dekodovat to naraz na pozadi

Ale nechcel som to komplikovat, ked to asi nebude potrebne.